File: python/array_and_hashing/560SubarraySumEqualsK.py
from typing import *


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        sum_freq = {0: 1}  # Initialize with sum 0 and frequency 1
        curr_sum = 0

        for num in nums:
            curr_sum += num
            # Check if there exists a subarray with sum equal to k
            if curr_sum - k in sum_freq:
                count += sum_freq[curr_sum - k]
            # Update the frequency of current sum
            sum_freq[curr_sum] = sum_freq.get(curr_sum, 0) + 1

        return count


sol = Solution()
print(sol.subarraySum([1, 2, 3], 3))

File: algo_zenith/stl/560SubarraySumEqualsK.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        map<int, int> mp;
        mp[0] = 1;
        int n = nums.size();
        int prefixSum = 0;
        int ans = 0;
        for(int i = 0; i < n; i++){
            prefixSum += nums[i];
            int temp = prefixSum - k;
            ans += mp[temp];
            mp[prefixSum]++;
        }
        return ans;
    }
};
int main(){
    vector<int> arr = {1, -1, 0, 0};
    int k = 0;
    Solution sol;
    cout << "Ans is "<< sol.subarraySum(arr, k);
    return 0;
}